### ISSUE TITLE
model.fit and model.fit_generator provide  difference result

### ISSUE THREAD

**Issue by beebrain** [keras-team/keras#9499]
I tried to convert my dataset to generator function. 
I plane using a big data. So create generator function for reading some data from the dataset. 
Now I create the model and train the model with a small dataset. It pretty good result the accuracy around 90-95%. However, I tried to use the same dataset with generator function but it provided a different result. 

```
batch_size = 10
num_classes = 25
epochs_step = 10
img_rows, img_cols = 128, 128


ds_gen = FA.File_Helper()
#  loading the data
(x_train, y_train,x_test,y_test) = ds_gen.unpackData(size=128)
# #
print(x_train.shape)  #(11250,128,128,3)
print(y_train.shape) #(11250,)
x_train = x_train.astype('float32')
x_test = x_test.astype('float32')
x_train /= 255
x_test /=255
y_train -=1   #reindex to 0
y_test -=1    #reindex to 0

input_shape = (img_rows, img_cols, 3)
print(input_shape)  #(11250,128,128,3)

# # #  the model
model = pretrained_model(input_shape,num_classes)   #loss='sparse_categorical_crossentropy',

weigth_file="./weights/PreTrain_weights-improvement-{epoch:02d}-{val_acc:.4f}.hdf5"
checkpoint = ModelCheckpoint(weigth_file, monitor='val_acc', verbose=1, save_best_only=True, mode='max')
tbCallBack = TensorBoard(log_dir='./Graph/Pretrain_128Px', histogram_freq=0, write_graph=True, write_images=True)
callbacks_list = [checkpoint,tbCallBack]

model.fit(x_train,y_train,
              batch_size=10,
              initial_epoch=0,
              epochs=100,
              verbose=1,
              validation_data=(x_test,y_test),
              callbacks = callbacks_list)

```
This code provided acc -> 90-95%.

**Comment by beebrain** [comment]
This my code for model.fit_generator
```
batch_size = 10
num_classes = 25
epochs_step = 10
img_rows, img_cols = 128, 128


def data_genArray(batch_size=10):
    ds_gen = FA.File_Helper()
    # # # # loading the data
    (x_train, y_train) = ds_gen.unpackDataTrain(size=128)
    x_train = x_train.astype('float32')
    x_train /= 255
    y_train -=1

    input_batch = np.zeros((batch_size, img_rows,img_cols,3),dtype=np.uint8)
    target_batch = np.zeros((batch_size, ),dtype=np.uint8)

    while True:
        for key in range(y_train.shape[0]):
            image = x_train[key]
            label = y_train[key]
            wrapped = key % batch_size
            input_batch[wrapped] = image  # rescale to 0-1
            target_batch[wrapped] = label
            if wrapped == batch_size - 1:
                yield input_batch.astype("float32"), target_batch

def data_genArrayTest(batch_size = 10):

    ds_gen = FA.File_Helper()
    # # # # loading the data
    (x_test, y_test) = ds_gen.unpackDataTest(size=128)
    x_test = x_test.astype('float32')
    x_test /= 255
    y_test -= 1

    x_test = x_test.reshape(x_test.shape[0], img_rows, img_cols, 3)
    input_batch = np.zeros((batch_size, img_rows, img_cols, 3), dtype=np.uint8)
    target_batch = np.zeros((batch_size,), dtype=np.uint8)

    while True:
        for key in range(y_test.shape[0]):
            image = x_test[key]
            label = y_test[key]

            wrapped = key % batch_size
            input_batch[wrapped] = image  # rescale to 0-1
            target_batch[wrapped] = label
            if wrapped == batch_size - 1:

                yield input_batch.astype("float32"), target_batch

model = pretrained_model(input_shape,num_classes)

weigth_file="./weights/PreTrain_weights-improvement-{epoch:02d}-{val_acc:.4f}.hdf5"
checkpoint = ModelCheckpoint(weigth_file, monitor='val_acc', verbose=1, save_best_only=True, mode='max')
tbCallBack = TensorBoard(log_dir='./Graph/Pretrain_128Px', histogram_freq=0, write_graph=True, write_images=True)
callbacks_list = [checkpoint,tbCallBack]

model.fit_generator(data_genArray(batch_size),
                    steps_per_epoch=11250/batch_size,
                    epochs=100,
                    validation_data=data_genArrayTest(batch_size),
                    validation_steps=3750/batch_size,
                    verbose=1,
                    callbacks=callbacks_list)
```
This code provided acc -> 50-60%

**Comment by hungbie** [comment]
Hi,

Seems like you only train on 10 samples everytime. When you yield you always go from 0 to your batch size in your input. If you want to be able to train on everything, choose the input randomly inside your generator.
